### PR TITLE
[SPARK-24692][TESTS] Improvement FilterPushdownBenchmark

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?
Refer to the [`WideSchemaBenchmark`](https://github.com/apache/spark/blob/v2.3.1/sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/WideSchemaBenchmark.scala) update `FilterPushdownBenchmark`:
1. Write the result to `benchmarks/FilterPushdownBenchmark-results.txt` for easy maintenance.
2. Add more benchmark case: `StringStartsWith`, `Decimal`, `InSet -> InFilters` and `tinyint`.

## How was this patch tested?

manual tests

### CODE REVIEW

**Comment by wangyum** [comment]
cc @maropu

**Comment by wangyum** [comment]
@HyukjinKwon Can you merge this to master first? I would like to update the [Benchmark results](https://github.com/apache/spark/pull/21677/files#diff-c5c0bfc86983d5779269cf75da8ed645) of several other pushdown related PRs to the corresponding PR.

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/benchmarks/FilterPushdownBenchmark-results.txt [hunk 1]
```diff
@@ -0,0 +1,556 @@
+############################[ Pushdown for many distinct value case ]############################
```
Comment (HyukjinKwon) [inline sql/core/benchmarks/FilterPushdownBenchmark-results.txt]: @wangyum, can we mimic any other format? For example, when I do such thing, I usually copy a format from another. For example, how about those below (which I am kind of used to).

```
Pushdown for many distinct value case:

  Java HotSpot(TM) 64-Bit Server VM 1.8.0_151-b12 on Mac OS X 10.12.6
  Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz

  Select 0 string row (value IS NULL):     Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
  ------------------------------------------------------------------------------------------------
  Parquet Vectorized                            7928 / 8019          2.0         504.0       1.0X
```

```
========================================================================
Pushdown for many distinct value case
========================================================================

Java HotSpot(TM) 64-Bit Server VM 1.8.0_151-b12 on Mac OS X 10.12.6
Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz

Select 0 string row (value IS NULL):     Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
Parquet Vectorized                            7928 / 8019          2.0         504.0       1.0X
```

```
Pushdown for many distinct value case ...

Java HotSpot(TM) 64-Bit Server VM 1.8.0_151-b12 on Mac OS X 10.12.6
Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz

Select 0 string row (value IS NULL):     Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
Parquet Vectorized                            7928 / 8019          2.0         504.0       1.0X
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/FilterPushdownBenchmark.scala [hunk 2]
```diff
@@ -44,8 +49,33 @@ object FilterPushdownBenchmark {
...
+  override def beforeEach(td: TestData) {
+    super.beforeEach(td)
+    val testName = "[ " + td.name + " ]"
+    val halfLength = (97 - testName.length) / 2
+    val testHeader = (("".padTo(halfLength, '#') + testName).padTo(97, '#') + "\n").getBytes
+    out.write(testHeader)
+  }
+
+  override def afterAll() {
+    super.afterAll()
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/FilterPushdownBenchmark.scala]: nit:
```
try {
  out.close()
} finally {
  super.afterAll()
}
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/FilterPushdownBenchmark.scala [hunk 3]
```diff
@@ -427,16 +245,122 @@ object FilterPushdownBenchmark {
...
 
       withTempTable("orcTable", "patquetTable") {
         prepareStringDictTable(dir, numRows, numDistinctValues, width)
-        runStringBenchmark(numRows, width, mid, "distinct string")
+        runStringBenchmark(numRows, width, numDistinctValues / 2, "distinct string")
+      }
+    }
+  }
+
+  ignore("Pushdown benchmark for StringStartsWith") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/FilterPushdownBenchmark.scala]: So those four below is the newly added benchmarks?

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise.

File: sql/core/benchmarks/FilterPushdownBenchmark-results.txt (hunk 1 above)
Comment (wangyum) [reply sql/core/benchmarks/FilterPushdownBenchmark-results.txt]: How about this?
```
...
Select all int rows (value != -1):       Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
Parquet Vectorized                            1140 / 1165          0.9        1087.4       1.0X
Parquet Vectorized (Pushdown)                 1140 / 1172          0.9        1086.8       1.0X
Native ORC Vectorized                         1158 / 1206          0.9        1104.7       1.0X
Native ORC Vectorized (Pushdown)              1151 / 1220          0.9        1098.1       1.0X


================================================================================================
Pushdown for few distinct value case (use dictionary encoding)
================================================================================================

Java HotSpot(TM) 64-Bit Server VM 1.8.0_151-b12 on Mac OS X 10.12.6
Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz

Select 0 distinct string row (value IS NULL): Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
Parquet Vectorized                             512 /  565          2.0         488.6       1.0X
Parquet Vectorized (Pushdown)                   27 /   33         39.3          25.5      19.2X
Native ORC Vectorized                          509 /  546          2.1         485.0       1.0X
Native ORC Vectorized (Pushdown)                79 /   91         13.2          75.5       6.5X
...
```

Comment (HyukjinKwon) [reply sql/core/benchmarks/FilterPushdownBenchmark-results.txt]: Yup, to be looks better. Let's go this way and correct it if any other committers have other preferences.

Comment (HyukjinKwon) [reply sql/core/benchmarks/FilterPushdownBenchmark-results.txt]: BTW that format was from our testing script as you might already know :-).

Comment (wangyum) [reply sql/core/benchmarks/FilterPushdownBenchmark-results.txt]: OK, I rebuild the benchmark use this format.

**Comment by HyukjinKwon** [comment]
I am merging this to show up other benchmark results in @wangyum's PRs.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by dongjoon-hyun** [comment]
Hi, @cloud-fan . 

As your comment https://github.com/apache/spark/pull/22336#issuecomment-418604019, I've came to this PR.

To be short, this PR changed the original `FilterPushdownBenchmark` to the current test-suite style one intentionally in order to follow `WideSchemaBenchmark` style. Now, I understand the reason why @wangyum file [SPARK-25339](https://issues.apache.org/jira/browse/SPARK-25339), but he cannot start for it.  :)

Could you confirm once more to revert this change? Also, do we want to change `WideSchemaBenchmark` together to be consistent for all benchmark suite?

**Comment by cloud-fan** [comment]
I need more context here. What's the benefit of the test suite style benchmark? I've committed benchmark code several times and I always use the main-method style. One benefit of the main-method style is, the benchmark probably can be more precise, without the overhead of the scalatest framework.

**Comment by dongjoon-hyun** [comment]
Me, too. I always used main-method style with the same reason. And many other BM are main-method style. According to this PR description, the reasons seems to be
1. There was a previous benchmark in that style; WideSchemaBenchmark
2. The result is recorded in the text file outside the code.

@wangyum . Do you want to add more?

If there is no other reasons, I'll start to rollback this one and convert `WideSchemaBenchmark` together in order to remove future confusions in dev community.

**Comment by dongjoon-hyun** [comment]
@wangyum . If you are still interested in reverting your PR as you mentioned in [SPARK-25339](https://issues.apache.org/jira/browse/SPARK-25339), please comment here about your thoughts and let us know. I believe that you are the best person to revert this.

**Comment by cloud-fan** [comment]
Seems we manually write benchmark result to a file, which can also be done with the main-method style.

**Comment by dongjoon-hyun** [comment]
Yes. @cloud-fan . We can embrace that concept to all the other main-method style benchmark. Previously, we do the manual copy&paste to put the result into the nearest place to the corresponding BM code. It's not an easy way for automation. 

With @wangyum 's that specific contribution, we can automate all benchmarks. Possibly, we can use that in the release process, too. So, are you heading `main-method` style with separate BM output files? For me, +1.

**Comment by cloud-fan** [comment]
> So, are you heading main-method style with separate BM output files?

Yes. So it's not reverting this PR, since writing BM result to a file is good. But we should update these BMs to use main-method style.

**Comment by dongjoon-hyun** [comment]
Thank you for the confirmation, @cloud-fan. Yes, Right. It's not reverting this PR. `main-method style with separate BM output` will be the standard style for all Spark benchmark code eventually from now.

So, @wangyum, are you available for this changes (starting from this suite)?

**Pull request closed without merging**